src/Pay/Adapter.php

```diff
@@ -0,0 +1,167 @@
+
+abstract class Adapter
+{
+
+  /**
+   * @var bool
+   */
+  protected bool $testMode;
+
+  /**
+   * @var string
+   */
+  protected string $currency;
+
+  /**
+   * Set test mode
+   */
+  public function setTestMode(bool $testMode)
+  {
+    $this->testMode = $testMode;
+  }
+
+  /**
+   * Get whether it's in test mode
+   */
+  public function getTestMode(): bool
+  {
+    return $this->testMode;
+  }
+
+  /**
+   * Get name of the payment gateway
+   */
+  abstract public function getName(): string;
+
+  /**
+   * Set the currency for payments
+   */
+  public function setCurrency(string $currency)
+  {
+    $this->currency = $currency;
+  }
+
+  /**
+   * Get currently set currency for payments
+   */
+  public function getCurrency(): string
+  {
+    return $this->currency;
+  }
+
+  /**
+   * Make a purchase request
+   */
+  abstract public function purchase(int $amount, string $customerId, string $cardId, array $additionalParams = []): array;
+
+  /**
+   * Refund payment
+   */
+  abstract public function refund(string $paymentId, int $amount = null): array;
+
+  /**
+   * Add a credit card for a customer
+   */
+  abstract public function createCard(string $customerId, string $cardId): array;
+
+  /**
+   * Update credit card
+   */
+  abstract public function updateCard(string $customerId, string $cardId, string $name = null, int $expMonth = null, int $expYear = null, array $billingDetails = null): array;
+
+  /**
+   * Get credit card
+   */
+  abstract public function getCard(string $customerId, string $cardId): array;
+
+  /**
+   * List cards
+   */
+  abstract public function listCards(string $customerId): array;
+
+  /**
+   * Remove a credit card for a customer
+   */
+  abstract public function deleteCard(string $customerId, string $cardId): bool;
+
+  /**
+   * Add new customer in the gateway database
+   * returns the id of the newly created customer
+   * 
+   * @throws Exception
+   */
+  abstract public function createCustomer(string $name, string $email, array $billingDetails = [], string $paymentMethod = null): array;
+
+  /**
+   * List customers
+   */
+  abstract public function listCustomers(): array;
+
+  /**
+   * Get customer details by ID
+   */
+  abstract public function getCustomer(string $customerId): array;
+
+  /**
+   * Update customer details
+   */
+  abstract public function updateCustomer(string $customerId, string $name, string $email, array $billingDetails = [], string $paymentMethod): array;
+
+  /**
+   * Delete customer by ID
+   */
+  abstract public function deleteCustomer(string $customerId): bool;
+
+  protected function call(string $method, string $url, mixed $body, array $headers = [], array $options = []): mixed
+  {
+
+    $responseHeaders = [];
+    $ch = \curl_init();
+
+    // define options
+    $optArray = array_merge(array(
+      CURLOPT_URL => $url,
+      CURLOPT_RETURNTRANSFER => true,
+      CURLOPT_CUSTOMREQUEST => $method,
+      CURLOPT_POSTFIELDS => $body,
```

Comment:
This will only work for POST requests right ? 
Since the call function is generic and supposed to serve more use-cases in the future, we should also handle parameters in case of GET requests. 

You can refer to the call method in the PHP SDK.

Reply:
https://github.com/appwrite/sdk-for-php/blob/00a81dd6dafccf94aa183749f286f559af5fabbe/src/Appwrite/Client.php#L169-L187